src/usp_generator.rs

```diff
@@ -144,6 +114,109 @@ pub fn usp_simple_error<'a>(code: u32, message: Option<String>) -> Body<'a> {
     }
 }
 
+/// Creates a body for a USP Msg with an USP Error
+///
+/// # Arguments
+///
+/// * `code` - The USP error code, MUST be between 7000 and 7999
+/// * `message` - An `Option<String>` containing the user readable message. Will be automatically
+///               filled in for standard error codes if not supplied
+/// * `param_errs` - A slice of a parameter path, error code and error message, can be empty
+///
+/// # Panics
+///
+/// Panics if both `code` is an invalid USP error code and `message` is `None`
+///
+/// # Examples
+///
+/// ```
+/// use rusp::usp_generator::usp_error;
+/// let err = usp_error::<&'static str>(7001, None, &[]);
+/// ```
+///
+/// ```
+/// use rusp::usp_generator::usp_error;
+/// let err = usp_error(
+///     7803,
+///     Some("Funny custom vendor error".into()),
+///     &[("Some.Path", 7804, "Funny error related to path")],
+/// );
+/// ```
+///
+/// ```should_panic
+/// use rusp::usp_generator::usp_error;
+/// let err = usp_error::<&'static str>(8000, None, &[]);
+/// ```
+pub fn usp_error<S: AsRef<str>>(
+    code: u32,
+    message: Option<String>,
+    param_errs: &[(S, u32, S)],
+) -> Body {
+    use crate::usp::mod_Body::OneOfmsg_body::*;
+    use crate::usp::mod_Error::ParamError;
+
+    let err_msg = message.unwrap_or_else(|| get_err_msg(code).to_string());
+
+    Body {
+        msg_body: error({
+            Error {
+                err_code: code,
+                err_msg: err_msg.into(),
+                param_errs: param_errs
+                    .iter()
+                    .map(|(param_path, err_code, err_msg)| ParamError {
+                        param_path: Cow::Borrowed(param_path.as_ref()),
+                        err_code: *err_code,
+                        err_msg: Cow::Borrowed(err_msg.as_ref()),
+                    })
+                    .collect(),
+            }
+        }),
+    }
+}
+
+/// Gets an USP error message from the error code
+pub fn get_err_msg(code: u32) -> &'static str {
```

Comment:
I wonder whether it would make sense to turn this into:

```suggestion
pub fn get_err_msg(code: u32) -> Option<&'static str> {
```

instead of panicking straight away...

Reply:
It does make sense, especially because I made that a public function. I will update it now.